Plumb graph_op through TFPartition, device partitioning, and graph lowering

lattner opened the pull request:
This required some minor changes to the device partitioning code, but overall
this is soo much simpler than the code it was will eventually replace.

The next step is to make TFPartition and device sends start unconditionally
using graph_op instructions. After that I'll do input lists, and then do
constexpr arrays.

lattner committed:
Plumb graph_op through TFPartition, device partitioning, and graph lowering.

This required some minor changes to the device partitioning code, but overall
this is soo much simpler than the code it was will eventually replace.

The next step is to make TFPartition and device sends start unconditionally
using graph_op instructions. After that I'll do input lists, and then do
constexpr arrays.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

rxwei commented on lib/SILOptimizer/Mandatory/TFUtilities.cpp:
```diff
@@ -1510,6 +1519,129 @@ SILInstruction *SILTensorOpInfo::canonicalizeOperands(
+      deviceType = DeviceType::GPU;
+  }
+
+  if (firstFound) {
```
Why not use an `else` to follow the last statement?

lattner replied:
That was existing code, I just moved it, fixed though, thanks!

lattner committed:
Tidy up, per review comment.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

mhong commented on lib/SILOptimizer/Mandatory/TFUtilities.cpp:
```diff
@@ -1510,6 +1519,126 @@ SILInstruction *SILTensorOpInfo::canonicalizeOperands(
+/// Scan the specified function, looking for logic that configures the current
+/// graph.
+GraphGlobalConfiguration
+GraphGlobalConfiguration::getForFunction(SILFunction &fn) {
```
it's great that we are extending the config from global to per-function. Some suggestions/questions:

1. should we rename GraphGlobalConfiguration to sth like DeviceConfiguration or DevicePlacementConfiguration? The name "global" is no longer appropriate.

2. should we extend the unit testing on TensorFlow.enableGPU() / enableTPU() for the per-function case? e.g. if foo() calls bar(), and we see such a statement in both foo() and bar(), i guess the one in bar() should error out, when we are partitioning foo(), and bar() is inlined into foo() by our deabstraction? 

It'd be fine to leave a TODO for #2 in this PR.

mhong reviewed (changes requested):
Great progress in using the new graph op inst! Left some minor/style-related comments. Will continue the review later.

mhong commented on lib/SILOptimizer/Mandatory/TFDevicePartition.cpp:
```diff
@@ -734,7 +781,7 @@ class DevicePartitionerImpl
     if (isa<SILArgument>(opValue)) return;
 
-    auto *operandInst = cast<SILInstruction>((SILNode *)opValue);
+    auto *operandInst = opValue->getDefiningInstruction();
```
add assert(operandInst)?

lattner replied:
Ok done.

mhong commented on lib/SILOptimizer/Mandatory/TFDevicePartition.cpp:
```diff
@@ -255,14 +256,49 @@ class DevicePartitionCloner
+    resultTypes.push_back(r->getType());
+
+  auto newOp = B.createGraphOperation(loc, inst->getName(), args,
+                                      inst->getAttributes(), resultTypes);
```
the use of "inst->getAttributes()" has an awesome pay-off in simplicity, compared to the tfop builtin!

lattner replied:
Yes, it's a huge improvement!

mhong commented:
Done with the review. Look forward to the full migration over to graph op inst!

lattner committed:
Fix an issue where GraphGlobalConfiguration::getForFunction would invalidate
pointrers in deabstraction by deleting ops.

lattner commented on test/TensorFlow/deabstraction_finished.swift:
```diff
@@ -30,11 +58,11 @@ public func f(a: Tensor<Float>, idx: Tensor<Int32>) -> Tensor<Float> {
 public func tensorShape() -> Tensor<Float> {
   let shape : TensorShape = [2]
-  // expected-error @+1 {{attribute 'value' requires a constant argument}}
+  // xpected-error @+1 {{attribute 'value' requires a constant argument}}
```
This is disabled.  I will look to re-enable this in the future.

lattner commented on lib/SILOptimizer/Mandatory/TFUtilities.h:
```diff
@@ -189,13 +192,11 @@ struct GraphGlobalConfiguration {
-  DeviceType chooseDevice(StringRef opType, StringRef opDevice) const {
-    if (!opDevice.empty()) return OpDeviceType(opDevice);
-
+  DeviceType chooseDevice(StringRef opType) const {
```
It is used by deabstraction.  When the migration is done, a lot of cleanup can happen as this stuff moves to deabstraction only.

lattner commented on lib/SILOptimizer/Mandatory/TFUtilities.h:
```diff
@@ -377,6 +387,45 @@ struct GraphGlobalConfiguration {
     static SILInstruction *decodeTensorFromScalarsND(ApplyInst *inst);
   };
 
+  /// Represent information about a TensorFlow operation as represented in SIL
```
Fixed, thanks!

lattner commented on lib/SILOptimizer/Mandatory/TFUtilities.cpp:
```diff
@@ -1510,6 +1519,126 @@ SILInstruction *SILTensorOpInfo::canonicalizeOperands(
+  return opName;
+}
+
+/// Given an attribute name like foo$dtype, decode the name and the class.
```
Fixed thanks!

lattner committed:
Improvements from code review feedback.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
Seriously.

@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

rxwei commented:
CI is stuck on me too.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
oh yeah :-)

lattner committed:
Split out errors detected in deabstraction into their own testfile,
since an error detected in deabstraction prevents partitioning from being run at all.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner committed:
For some reason I'm getting a new pattern for zero element arrays being
fed into the __tf_tensor_from_scalars_1d runtime function call.  This was
unpredictably causing a new send/receive which crashed graph lowering.
Fix this by recognizing the new pattern.

Deabstraction should define this issue away when it can do arrays... soon.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow
